Create backups when writing files (SOFTWARE-2352)

brianhlin opened the pull request:
https://jira.opensciencegrid.org/browse/SOFTWARE-2352

I've thought about changing the `<path> + '.new'` invocations for the various openssl outputs to using `tempfile.NamedTemporaryFile` but the fd is left open so I think openssl would fail when trying to write output to the named file. On the other hand we'd lose atomicity with `tempfile.mktemp`.

brianhlin committed:
Create backups when writing files (SOFTWARE-2352)

bbockelm commented:
I think OpenSSL would only fail if it also tried to do an atomic write.  The fact that the calling process has an open FD shouldn't matter.

(Can OpenSSL write to stdout?  If so, you can hook stdout to a pipe and have python actually write to the named temporary file...)

matyasselmeci commented on lib/cagen.py:
```diff
@@ -125,30 +129,32 @@ def usercert(self, username, password, days=None):
         user_request = 'user_req'
 
         try:
             os.makedirs(globus_dir, 0755)
```
For future-proofing, consider changing octal constants from `0XXX` to `0oXXX`. (The former syntax is not supported in py3.)

matyasselmeci commented on lib/cagen.py:
```diff
@@ -125,30 +129,32 @@ def usercert(self, username, password, days=None):
         try:
             os.makedirs(globus_dir, 0755)
+            os.chown(globus_dir, user.pw_uid, user.pw_gid)
         except OSError, e:
```
For future-proofing, use `except OSError as e`

matyasselmeci commented on lib/cagen.py:
```diff
@@ -309,8 +317,50 @@ def _get_hostname():
+        uid = stat.st_uid
+        gid = stat.st_gid
+        os.rename(target_path, target_path + '.old')
+    except IOError, exc:
```
File not found raises `OSError`, not `IOError`. Also see above quibble about exception syntax.

brianhlin replied:
Opening a non-existent file actually does raise IOError. Should I catch `IOError`, and `OSError`?

matyasselmeci replied:
Interesting! The Python manual is a little crappy on the subject. In this case, you should change the except statement in `usercert()` above.

To be honest, I never remember which problems are IOError and which ones are OSError, so I always catch their base class, EnvironmentError.

brianhlin replied:
[This one](https://github.com/opensciencegrid/osg-ca-generator/pull/7/files#diff-aa56a9951831a6f2695c73c36d214137R134)? `os.makedirs` and `os.chown` raise `OSError` if the folder already exists or if the folder doesn't exist, respectively.

matyasselmeci replied:
So for nonexistent files it's IOError and for nonexistent directories it's OSError? This is terrible.

Apparently in Python 3, IOError = OSError = EnvironmentError, so the Python devs agree with me.

brianhlin replied:
I think it depends on the function:

* `open()` on a non-existent file throws IOError
* `os.chown()` on a non-existent file OR directory throws OSError

Regardless, I can change them all to EnvironmentError and be done with it

matyasselmeci replied:
If you're sure it works then you don't need to, but it would make me more comfortable if you did.

matyasselmeci reviewed (dismissed):
One error plus a few syntax quibbles.

brianhlin committed:
Fixup issues raised from PR

brianhlin committed:
Move chown/chmod outside of _safe_move

Let the caller be responsible for setting permissions/ownership
on the new file properly

brianhlin committed:
Grab openssl output from stdout or write them to temp files

brianhlin commented:
@matyasselmeci this should be ready for re-review. I've tested the changes on a fermicloud VM successfully

brianhlin committed:
Use EnvironmentError, the base class of OSError/IOError

Some functions raise IOError, some raise OSError when opening
non-existent files. Just use EnvironmentError.

Also, RuntimeError for unexpected output for retrieving a cert's
subject/issuer is more appropriate than OSError
